torch/csrc/distributed/c10d/c10d_frontend.cpp

```diff
@@ -0,0 +1,635 @@
+  } else {
+    int64_t group_dst_rank = getGroupRank(group, dst);
+    opts.rootRank = group_dst_rank;
+    work = group->reduce(tensor_list, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::reduce(
+    at::Tensor tensor,
+    int64_t dst,
+    const std::shared_ptr<ProcessGroup>& group,
+    ReduceOp op,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  ReduceOptions opts;
+  opts.reduceOp = op;
+  opts.rootRank = dst;
+
+  checkDefaultPg();
+  std::shared_ptr<ProcessGroup::Work> work;
+  std::vector<at::Tensor> tensors = {std::move(tensor)};
+  if (group == default_pg_) {
+    work = group->reduce(tensors, opts);
+  } else {
+    int64_t group_dst_rank = getGroupRank(group, dst);
+    opts.rootRank = group_dst_rank;
+    work = group->reduce(tensors, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGatherMultiGPU(
+    std::vector<std::vector<at::Tensor>>& output_tensor_lists,
+    std::vector<at::Tensor>& input_tensor_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto work = group->allgather(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGather(
+    std::vector<at::Tensor>& tensor_list,
+    at::Tensor tensor,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  std::vector<std::vector<at::Tensor>> output_tensor_lists = {std::move(tensor_list)};
+  std::vector<at::Tensor> input_tensor_list = {std::move(tensor)};
+  auto work = group->allgather(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGatherCoalesced(
+    std::vector<std::vector<at::Tensor>>& output_tensor_lists,
+    std::vector<at::Tensor>& input_tensor_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto work =
+      group->allgather_coalesced(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::gather(
+    at::Tensor tensor,
+    const std::vector<at::Tensor>& gather_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    int64_t dst,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto my_rank = getRank(default_pg_);
+
+  // TODO TODO how to implement this?
+  // _validate_output_list_for_rank(my_rank, dst, gather_list)
```

Comment:
IIUC, `my_rank == group->getRank()`. Could you please elaborate the problem that prevents us from implementing `_validate_output_list_for_rank`?

Reply:
Changed line 462 to `my_rank = group->getRank()`.

What's preventing me from implementing _validate_output_list_for_rank is that now `gather_list` is not null-able. Unsure what the erroring behavior should be. Or should it really be nullable?

Reply:
I see. Good point. `gather_list` from Python API can `None`. Does it help if we use a `shared_ptr` instead of a reference?